redwood-lazylayout-widget/src/commonMain/kotlin/app/cash/redwood/lazylayout/widget/LazyListScrollProcessor.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2023 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.redwood.lazylayout.widget
+
+import app.cash.redwood.lazylayout.api.ScrollItemIndex
+
+public abstract class LazyListScrollProcessor {
+  /** To notify guest code of user scrolls. */
+  private var onViewportChanged: ((Int, Int) -> Unit)? = null
+
+  /** We can't scroll to this index until we have enough data for it to display! */
+  private var deferredProgrammaticScrollIndex: Int = -1
+
+  /** Once we receive a user scroll, we stop forwarding programmatic scrolls. */
+  private var userHasScrolled = false
+
+  public fun onViewportChanged(onViewportChanged: (Int, Int) -> Unit) {
+    this.onViewportChanged = onViewportChanged
+  }
+
+  public fun scrollItemIndex(scrollItemIndex: ScrollItemIndex) {
+    // Defer until we have data in onEndChanges().
+    deferredProgrammaticScrollIndex = scrollItemIndex.index
+  }
+
+  public fun onEndChanges() {
+    // Do nothing: we don't have deferred scrolls.
+    if (deferredProgrammaticScrollIndex == -1) return
+
+    // Do nothing: we don't do programmatic scrolls if the user has already scrolled manually.
+    if (userHasScrolled) return
+
+    // Do nothing: we can't scroll to this item because it hasn't loaded yet!
+    if (contentSize() <= deferredProgrammaticScrollIndex) return
```

Comment:
This is also important  . . . we defer applying programmatic scrolls until there’s enough rows!

Reply:
I'm a bit worried about keeping the deferred value basically forever. Should we clear unconditionally after first non-zero size even if it's insufficient items (such as if data changed in background reducing rows)? This would depend on us loading the target range in a single operation rather than requiring multiple.

Reply:
If we have fewer rows when restoring (for whatever reason), and then later we gain some additional rows such that the index suddenly becomes valid, would it trigger a sudden scroll? Even if the user had done some manual scrolling?

Reply:
Okay I should have read one method down... 🤦 . I think these are covered!